pkg/windows/calico.go

```diff
@@ -52,147 +53,226 @@ const (
+	return nil
 }
 
-// createCNIConfig creates the CNI config for the CNI.
-func createCNIConfig(config *CNIConfig) error {
-	parsedTemplate, err := parseTemplateFromConfig(calicoConfigTemplate, config)
+// overrideCalicoConfigByHelm overrides the default values set for calico if a Chart exists
+func (c *Calico) overrideCalicoConfigByHelm(restConfig *rest.Config) error {
+	hc, err := helm.NewFactoryFromConfig(restConfig)
 	if err != nil {
 		return err
 	}
-	return util2.WriteFile(filepath.Join(config.NodeConfig.AgentConfig.CNIConfDir, CalicoConfigName), parsedTemplate)
-}
 
-// getDefaultConfig sets the default configuration.
-func getDefaultConfig(config *CNIConfig, dataDir string, nodeConfig *config.Node) error {
-	platformType, err := getPlatformType()
+	cniChartConfig, err := hc.Helm().V1().HelmChartConfig().Get(metav1.NamespaceSystem, CalicoChart, metav1.GetOptions{})
 	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return fmt.Errorf("failed to check for %s HelmChartConfig", CalicoChart)
+	}
+	if cniChartConfig == nil {
+		return nil
+	}
+	overrides := CalicoInstallation{}
+	if err := yaml.Unmarshal([]byte(cniChartConfig.Spec.ValuesContent), &overrides); err != nil {
 		return err
 	}
-
-	calicoCfg := &CalicoConfig{
-		Name:                  "Calico",
-		Hostname:              nodeConfig.AgentConfig.NodeName,
-		NodeNameFile:          filepath.Join("c:\\", dataDir, "agent", CalicoNodeNameFileName),
-		KubeNetwork:           "Calico.*",
-		Mode:                  "vxlan",
-		ServiceCIDR:           nodeConfig.AgentConfig.ServiceCIDR.String(),
-		DNSServers:            nodeConfig.AgentConfig.ClusterDNS.String(),
-		DNSSearch:             "svc." + nodeConfig.AgentConfig.ClusterDomain,
-		DatastoreType:         "kubernetes",
-		NetworkingBackend:     "vxlan",
-		Platform:              platformType,
-		StartUpValidIPTimeout: 90,
-		LogDir:                "",
-		IP:                    "autodetect",
-		IPAutoDetectionMethod: "first-found",
-		Felix: FelixConfig{
-			Metadataaddr:    "none",
-			Vxlanvni:        "4096",
-			MacPrefix:       "0E-2A",
-			LogSeverityFile: "none",
-			LogSeveritySys:  "none",
-		},
-		CNI: CalicoCNIConfig{
-			BinDir:   nodeConfig.AgentConfig.CNIBinDir,
-			ConfDir:  nodeConfig.AgentConfig.CNIConfDir,
-			IpamType: "calico-ipam",
-			Version:  "0.3.1",
-		},
+	// Marshal for clean debug logs, otherwise it's all pointers
+	b, err := yaml.Marshal(overrides)
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("calico override found: %s\n", string(b))
+	if nodeV4 := overrides.Installation.CalicoNetwork.NodeAddressAutodetectionV4; nodeV4 != nil {
+		c.CalicoCNICfg.CalicoConfig.IPAutoDetectionMethod = getNodeAddressAutodetection(*nodeV4)
 	}
-	config.CalicoConfig = calicoCfg
+
 	return nil
 }
 
+func getNodeAddressAutodetection(autoDetect opv1.NodeAddressAutodetection) string {
+	if autoDetect.FirstFound != nil && *autoDetect.FirstFound {
+		return "first-found"
+	} else if autoDetect.CanReach != "" {
```

Comment:
Can all of the same logic be had wihtout all of the "else" statements?

Reply:
In my defense, I must say that this is old code ;). Do you mean adding separate if conditions? As it evaluates different subfields of `autoDetect` we can't have a `switch` :(